odl/solvers/functional/default_functionals.py

```diff
@@ -0,0 +1,452 @@
+
+    The L1-norm, ``||x||_1``,  is defined as the integral/sum of ``|x|``.
+
+    Notes
+    -----
+    If the functional is defined on an :math:`\mathbb{R}^n`-like space, the
+    :math:`L_1`-norm is defined as
+
+    .. math::
+
+        \| x \|_1 = \\sum_{i = 1}^n |x_i|.
+
+    If the functional is defined on an :math:`L_2`-like space, the
+    :math:`L_1`-norm is defined as
+
+    .. math::
+
+        \| x \|_1 = \\int_{\Omega} |x| dx.
+
+    """
+
+    def __init__(self, space):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        """
+        super().__init__(space=space, linear=False, grad_lipschitz=np.inf)
+
+    def _call(self, x):
+        """Return the L1-norm of ``x``."""
+        return x.ufunc.absolute().inner(self.domain.one())
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        functional = self
+
+        class L1Gradient(Operator):
+
+            """The gradient operator of this functional."""
+
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Apply the gradient operator to the given point."""
+                return x.ufunc.sign()
+
+        return L1Gradient()
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        See Also
+        -----
+        proximal_l1 : proximal factory for the L1-norm.
+        """
+        return proximal_l1(space=self.domain)
+
+    @property
+    def convex_conj(self):
+        """The convex conjugate functional of the L1-norm."""
+        return IndicatorLpUnitBall(self.domain, exponent=np.inf)
+
+
+class IndicatorLpUnitBall(Functional):
+
+    """The indicator function on the unit ball in given the ``Lp`` norm.
+
+    It does not implement `gradient` since it is not differentiable everywhere.
+
+    Notes
+    -----
+    This functional is defined as
+
+        .. math::
+
+            f(x) = \\left\{ \\begin{array}{ll}
+            0 & \\text{if } ||x||_{L_p} \\leq 1, \\\\
+            \\infty & \\text{else}
+            \\end{array} \\right.
+
+    where :math:`||x||_{L_p}` is the :math:`L_p`-norm, and for
+    :math:`p = \\infty` it is defined as
+
+        .. math::
+
+            ||x||_{\\infty} = \max_x (|x|).
+    """
+
+    def __init__(self, space, exponent):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp` or `FnBase`
+            Domain of the functional.
+        exponent : int or infinity
+            Specifies wich norm to use.
+        """
+        super().__init__(space=space, linear=False)
+
+        if exponent is not np.inf and not isinstance(exponent, int):
```

Comment:
ufunc.power works for arbitrary exponents, frankly also for negative even though the cauchy swartz inequality fails there.